syntax = "proto3";

package pieledger;

enum AccountType {
  ROOT = 0;
  RECEIVABLE = 1;
  MUTUAL = 2;
  CASH = 3;
  ASSET = 4;
  BANK = 5;
  STOCK = 6;
  CREDIT = 7;
  LIABILITY = 8;
  PAYABLE = 9;
  INCOME = 10;
  EXPENSE = 11;
  TRADING = 12;
  EQUITY = 13;
}

message MonetaryAmount {
  string value = 1;
  sint64 num = 2;
  int32 denom = 3;
}

message Account {
  string guid = 1;
  string name = 2;
  AccountType type = 3;
  Account parent = 4;
  bool placeholder = 5;
  MonetaryAmount balance = 6;
}

enum SplitAction {
  UNKNOWN = 0;
  BUY = 1;
  SELL = 2;
  TOPPING = 3;
  CASHIN = 4;
  WITHDRAW = 5;
  CASHOUT = 6; 
  REFUND = 7;
}

message Split {
  string guid = 1;
  Account account = 2;
  SplitAction action = 3;
  MonetaryAmount amount = 4;
  string memo = 5;
  Transaction transaction = 6;
}

message Transaction {
  string guid = 1;
  string reference = 2;
  string description = 3;
  
  repeated Split splits = 4;
}
